File: app/src/main/java/de/kai_morich/simple_usb_terminal/BGapi.java
package de.kai_morich.simple_usb_terminal;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class BGapi {

    private static final HashMap<String, byte[]> knownResponses = new HashMap<>();
    private static final HashMap<String, String> commands = new HashMap<>();

    static {
        knownResponses.put("scanner_set_mode_rsp", new byte[]{0x20, 0x02, 0x05, 0x02, 0x00, 0x00});
        knownResponses.put("scanner_set_timing_rsp", new byte[]{0x20, 0x02, 0x05, 0x01, 0x00, 0x00});
        knownResponses.put("connection_set_default_parameters_rsp", new byte[]{0x20, 0x02, 0x06, 0x00, 0x00, 0x00});
        knownResponses.put("scanner_start_rsp", new byte[]{0x20, 0x02, 0x05, 0x03, 0x00, 0x00});
        knownResponses.put("scanner_stop_rsp", new byte[]{0x20, 0x02, 0x05, 0x05, 0x00, 0x00});
        knownResponses.put("message_rotate_cw_rsp", new byte[]{0x20, 0x04, (byte) 0xFF, 0x00, 0x00, 0x00, 0x01, 0x01});
        knownResponses.put("message_rotate_ccw_rsp", new byte[]{0x20, 0x04, (byte) 0xFF, 0x00, 0x00, 0x00, 0x01, 0x02});
        knownResponses.put("message_rotate_stop_rsp", new byte[]{0x20, 0x04, (byte) 0xFF, 0x00, 0x00, 0x00, 0x01, 0x03});

        commands.put("scanner_set_mode", "200205020401");
        commands.put("scanner_set_timing", "200505010410001000");
        commands.put("connection_set_parameters", "200c060050005000000064000000ffff");
        commands.put("scanner_start", "200205030402");
        commands.put("scanner_stop", "20000505");
        commands.put("message_rotate_cw", "2002FF000101");
        commands.put("message_rotate_ccw", "2002FF000102");
        commands.put("message_rotate_stop", "2002FF000103");
    }

    public static final String SCANNER_SET_MODE = commands.get("scanner_set_mode");
    public static final String SCANNER_SET_TIMING = commands.get("scanner_set_timing");
    public static final String CONNECTION_SET_PARAMETERS = commands.get("connection_set_parameters");
    public static final String SCANNER_START = commands.get("scanner_start");
    public static final String SCANNER_STOP = commands.get("scanner_stop");
    public static final String ROTATE_CW = commands.get("message_rotate_cw");
    public static final String ROTATE_CCW = commands.get("message_rotate_ccw");
    public static final String ROTATE_STOP = commands.get("message_rotate_stop");

    public static boolean isScanReportEvent(byte[] bytes) {
        // Note: the casting of 0xA1 is necessary because Java thinks that signed hex should exist?
        //      and as a result does funny things
        return bytes.length > 3 && bytes[0] == (byte) 0xA1 && bytes[1] == 0x00
                && bytes[2] == 0x05 && bytes[3] == 0x01;
    }

    public static boolean isKnownResponse(byte[] bytes) {
        for (byte[] response : knownResponses.values()) {
            if (Arrays.equals(response, bytes))
                return true;
        }
        return false;
    }

    public static String getResponseName(byte[] bytes) {
        for (Map.Entry<String, byte[]> entry : knownResponses.entrySet()) {
            if (Arrays.equals(entry.getValue(), bytes)) {
                return entry.getKey();
            }
        }
        return null;
    }

    public static String getCommandValue(String msg) {
        for (Map.Entry<String, String> entry : commands.entrySet()) {
            if (entry.getKey().equals(msg)) {
                return entry.getValue();
            }
        }
        return null;
    }

    public static String getCommandName(String msg) {
        for (Map.Entry<String, String> entry : commands.entrySet()) {
            if (entry.getValue().equals(msg)) {
                return entry.getKey();
            }
        }
        return null;
    }

    public static boolean isCommand(String msg) {
        for (Map.Entry<String, String> entry : commands.entrySet()) {
            if (entry.getValue().equals(msg)) {
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/kotlin/de/kai_morich/simple_usb_terminal/BGapi.kt
package de.kai_morich.simple_usb_terminal

/**
 * A wrapper class for the constants we use to communicate with the Gecko.
 *
 * While these technically should be findable in Silicon Lab's documentation, it's a lot easier
 *  to open up the NCP Commander in Simplicity Studio, manually send the commands there, and turn
 *  on the option to display the bytes that were sent
 * */
class BGapi {
    enum class Response (val bytes: ByteArray) {
        SCANNER_SET_MODE(byteArrayOf(0x20, 0x02, 0x05, 0x02, 0x00, 0x00)),
        SCANNER_SET_TIMING(byteArrayOf(0x20, 0x02, 0x05, 0x01, 0x00, 0x00)),
        CONNECTION_SET_DEFAULT_PARAMETERS(byteArrayOf(0x20, 0x02, 0x06, 0x00, 0x00, 0x00)),
        SCANNER_START(byteArrayOf(0x20, 0x02, 0x05, 0x03, 0x00, 0x00)),
        SCANNER_START_FAILED(byteArrayOf(0x20, 0x02, 0x05, 0x03, 0x02, 0x00)),
        SCANNER_STOP(byteArrayOf(0x20, 0x02, 0x05, 0x05, 0x00, 0x00)),
        MESSAGE_ROTATE_CW(byteArrayOf(0x20, 0x04, 0xFF.toByte(), 0x00, 0x00, 0x00, 0x01, 0x01)),
        MESSAGE_ROTATE_CCW(byteArrayOf(0x20, 0x04, 0xFF.toByte(), 0x00, 0x00, 0x00, 0x01, 0x02)),
        MESSAGE_ROTATE_STOP(byteArrayOf(0x20, 0x04, 0xFF.toByte(), 0x00, 0x00, 0x00, 0x01, 0x03)),
        MESSAGE_ROTATE_RELATED(byteArrayOf(0x20, 0x03, 0xFF.toByte(), 0x00, 0x01, 0x00, 0x00)),
        SYSTEM_BOOT(byteArrayOf(
            0xA0.toByte(), 0x12, 0x01, 0x00, 0x03, 0x00, 0x03, 0x00, 0x02, 0x00,
            0x96.toByte(), 0x01, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0xE0.toByte(), 0x5C, 0x35, 0x51))
        ;

        companion object {
            @JvmStatic
            fun getResponseName(bytes: ByteArray): String? {
                return values().find {it.bytes.contentEquals(bytes)}?.name
            }

            @JvmStatic
            fun isKnownResponse(bytes: ByteArray): Boolean {
                return values().any {it.bytes.contentEquals(bytes)}
            }
        }
    }

    enum class Command (val value: String) {
        SCANNER_SET_MODE("200205020401"),
        SCANNER_SET_TIMING("200505010410001000"),
        CONNECTION_SET_PARAMETERS("200c060050005000000064000000ffff"),
        SCANNER_START("200205030402"),
        SCANNER_STOP("20000505"),
        ROTATE_CW("2002FF000101"),
        ROTATE_CCW("2002FF000102"),
        ROTATE_STOP("2002FF000103"),
        GET_TEMP("2002FF000104");

        companion object {
            @JvmStatic
            fun getCommandName(msg: String): String? {
                return values().find {it.value.contentEquals(msg)}?.name
            }

            @JvmStatic
            fun isCommand(msg: String): Boolean {
                return values().any {it.value.contentEquals(msg)}
            }
        }

    }

    companion object {
        @JvmStatic
        fun isScanReportEvent(byteArray: ByteArray): Boolean {
            val magicBytes = byteArrayOf(0xA0.toByte(), 0x1B.toByte(), 0x05.toByte(), 0x02.toByte())
            return byteArray.size >= magicBytes.size && byteArray.copyOfRange(0, magicBytes.size).
                contentEquals(magicBytes)
        }

        @JvmStatic
        fun isTemperatureResponse(bytes: ByteArray): Boolean {
            return bytes.size == 9 && bytes.last() == 0x69.toByte()
        }

    }


}